src/mainwin.cc

```diff
@@ -2027,6 +2033,170 @@ void MainWindow::csgRender()
+	}
+	
+	//Upload the file to the 3D Printing server and get the corresponding url to see it.
+	//The result is put in partUrl.
+	try
+	{
+        uploadStlAndGetPartUrl(exportFilename, userFacingName, partUrl);
+    }
+    catch (std::exception & e)
+    {
+		PRINT(e.what());
+        return;
+    }
+    
+    setCurrentOutput();
+
+	//Open the url in the default browser:
+	QDesktopServices::openUrl ( partUrl );
+}
+
+//This function uploads an stl to the 3D printing API endpoint and returns a url that, 
+//when accessed, will show the stl file as a part that can be configured and added to the 
+//shopping cart.  If it's not successful, it throws an exception with a message.
+//
+//Inputs:
+//    exportFilename  - The path to the temporary file that has the stl export in it.
+//    userFacingName  - Then name we should give the file when it is uploaded for the order process.
+//Outputs:
+//    partUrl         - The resulting url to go to next to continue the order process.
+void MainWindow::uploadStlAndGetPartUrl(const QString & exportFilename, const QString &userFacingName, QUrl &partUrl)
+{
+	setCurrentOutput();
+	
+	//Create a request:
+	QNetworkRequest request(QUrl("https://print.openscad.org/api/v1/part-upload/"));
+	
+	//Set the content header:
+	request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
+	
+	//Get the file name from the file path:
+	QString fileNameBase=QFileInfo(exportFilename).fileName();
+	
+	//Create the request:
+	QJsonObject jsonInput;
+	
+	//Start building the json request:
+	jsonInput.insert("fileName", userFacingName);
+	
+	//Read our stl file:
+	QFile file(exportFilename);
+	if (!file.open(QIODevice::ReadOnly))
+	{
+		throw std::runtime_error("Unable to open exported stl file.");
+	}
+	
+	//Convert it to base 64:
+	QString stlFileB64=file.readAll().toBase64();
+	
+	//Base 64-encoded file contents:
+	jsonInput.insert("file", stlFileB64);
+    
+    //Make sure it's there:
+    if (jsonInput.value("file") == QJsonValue::Undefined)
+        throw std::runtime_error("Could not enode stl into JSON.  Perhapse it is too large of a file? Try simplifying.");
+        
+	PRINTD("Sending this JSON:");
+	PRINTD(QString(QJsonDocument(jsonInput).toJson()).toLocal8Bit().constData());
+	
+	//Create a network access manager:
+	QNetworkAccessManager nam;
+	
+	//Send the post:
+	QNetworkReply *reply = nam.post(request, QJsonDocument(jsonInput).toJson());
+
+	//Wait for the reply:
+	while(!reply->isFinished())
+	{
+		qApp->processEvents();
+	}
+	
+	QVariant statusCodeV = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute);  
+	
+	//Clean up the reply object:
+	reply->deleteLater();
+    
+    char errorMessage[150];
+	
+    //If it wasn't successful:
+	if (! (statusCodeV.toInt()==200 or statusCodeV.toInt()==201 ))
+	{
+        sprintf(errorMessage, "An error occured while contacting the 3D Print API.  \nStatus code returned: %d", statusCodeV.toInt());
```

Comment:
Just stating, meaning I have no preference either way:
PRINTB internally uses boost::format.
boost::format uses a dynamic stream, so you do not need the char buffer `char errorMessage[150];`.

As this is part of the GUI, QString::sprintf() and QString::arg() would also be option, but as the error message is a char[], using QString would be a bit overcomplicated.

Reply:
I added this to #2664.